File: packages/core/src/common/path-boundary.ts
import * as fs from "fs";
import * as path from "path";
import type { PermissionScope } from "../settings";
import { isAbsoluteFilePath, normalizeFilePath } from "./state";

// Pure path-boundary primitives with no permission semantics, so tool
// handlers can enforce the execution-time boundary gate without importing
// the permission engine (specs/sandbox/design.md §4.1). isPathInProject /
// safeRealPath / isPathInAnyDirectory moved verbatim from permissions.ts.

export function isPathInProject(projectRoot: string, filePath: string): boolean {
  const normalized = normalizeFilePath(filePath);
  const absolutePath = isAbsoluteFilePath(normalized) ? normalized : path.resolve(projectRoot, normalized);
  // Hardening (deep review 2026-08-15, B3): a symlink planted INSIDE the
  // project could point at /etc while lexical resolution still classifies the
  // target as in-project (and thus pre-allowed). When both ends exist, prefer
  // the real paths; fall back to the lexical answer when realpath fails
  // (file not created yet / permissions), preserving prior behavior.
  const realPath = safeRealPath(absolutePath);
  const realRoot = safeRealPath(projectRoot);
  if (realPath && realRoot) {
    const realRelative = path.relative(realRoot, realPath);
    if (realRelative === "" || (!realRelative.startsWith("..") && !path.isAbsolute(realRelative))) {
      return true;
    }
    // realpath resolved and clearly escapes — trust it over the lexical answer.
    if (realRelative.startsWith("..") || path.isAbsolute(realRelative)) {
      return false;
    }
  }
  const relative = path.relative(path.resolve(projectRoot), path.resolve(absolutePath));
  return relative === "" || (!relative.startsWith("..") && !path.isAbsolute(relative));
}

export function safeRealPath(target: string): string | null {
  try {
    return fs.realpathSync(target);
  } catch {
    return null;
  }
}

export function isPathInAnyDirectory(
  projectRoot: string,
  filePath: string,
  directories: string[] | undefined
): boolean {
  if (!directories?.length) {
    return false;
  }

  const normalized = normalizeFilePath(filePath);
  const absolutePath = isAbsoluteFilePath(normalized) ? normalized : path.resolve(projectRoot, normalized);
  for (const directory of directories) {
    const normalizedDirectory = normalizeFilePath(directory);
    const absoluteDirectory = isAbsoluteFilePath(normalizedDirectory)
      ? normalizedDirectory
      : path.resolve(projectRoot, normalizedDirectory);
    const relative = path.relative(path.resolve(absoluteDirectory), path.resolve(absolutePath));
    if (relative === "" || (!relative.startsWith("..") && !path.isAbsolute(relative))) {
      return true;
    }
  }
  return false;
}

// ---------------------------------------------------------------------------
// Execution-time path boundary gate (P0, specs/sandbox/design.md §4.1).
//
// The permission layer already classifies paths as in/out-of-cwd, but that
// decision never reached the handlers — after a pass they ran with full host
// authority (gap G1/G2). PathGrant is the capability handed down per tool
// call; the gates enforce it right before any fs mutation/read.

export type PathGrant = {
  /** Allowed write roots (realpath-normalized). Always contains realpath(projectRoot). */
  readonly writeRoots: readonly string[];
  /** Allowed read roots: realpath(projectRoot) + readPermissionExemptPaths. */
  readonly readRoots: readonly string[];
  /**
   * This call's write-out-cwd scope resolved to "allow" (explicit user grant
   * or one-time approval). Roots are the static boundary; the booleans are
   * the dynamic per-call authorization — a one-time approval cannot be
   * expressed as a root list ("the whole disk"), hence the orthogonal flag.
   */
  readonly allowWriteOutsideRoots: boolean;
  /** This call's read-out-cwd scope resolved to "allow". */
  readonly allowReadOutsideRoots: boolean;
};

export type GateVerdict = { ok: true } | { ok: false; reason: string; scope: PermissionScope };

/**
 * Write boundary gate. `projectRoot` is only consulted when `grant` is
 * undefined (direct handler invocations without the session plumbing): the
 * grant then degenerates to projectRoot-only with both outside-roots flags
 * false — fail-closed for out-of-project writes, ordinary in-project work
 * unaffected. Without any root at all, everything is denied.
 */
export function gateWrite(grant: PathGrant | undefined, filePath: string, projectRoot?: string): GateVerdict {
  const candidate = resolveGateCandidate(filePath);
  const roots = grant ? grant.writeRoots : projectRoot ? [projectRoot] : [];
  if (isPathInRoots(roots, candidate)) {
    return { ok: true };
  }
  if (grant?.allowWriteOutsideRoots) {
    return { ok: true };
  }
  return {
    ok: false,
    scope: "write-out-cwd",
    reason: `Write target is outside the allowed write boundary: ${filePath}. This path was not authorized for out-of-project writes. If writing there is genuinely required, ask the user to grant the "write-out-cwd" permission.`,
  };
}

/** Read boundary gate, mirror semantics of {@link gateWrite}. */
export function gateRead(grant: PathGrant | undefined, filePath: string, projectRoot?: string): GateVerdict {
  const candidate = resolveGateCandidate(filePath);
  const roots = grant ? grant.readRoots : projectRoot ? [projectRoot] : [];
  if (isPathInRoots(roots, candidate)) {
    return { ok: true };
  }
  if (grant?.allowReadOutsideRoots) {
    return { ok: true };
  }
  return {
    ok: false,
    scope: "read-out-cwd",
    reason: `Read target is outside the allowed read boundary: ${filePath}. This path was not authorized for out-of-project reads. If reading there is genuinely required, ask the user to grant the "read-out-cwd" permission.`,
  };
}

const MAX_SYMLINK_DEPTH = 10;

/**
 * Chase the final symlink chain manually (realpath would fail on a dangling
 * link and drop us back to the lexical link path, which would let a
 * `<root>/link -> /etc/new-file` write escape the boundary). Depth-capped to
 * break cycles; a chain over the cap keeps its lexical tail and simply fails
 * containment against outside roots.
 */
function followSymlinkChain(target: string, depth: number): string {
  if (depth >= MAX_SYMLINK_DEPTH) {
    return target;
  }
  let stat: fs.Stats;
  try {
    stat = fs.lstatSync(target);
  } catch {
    return target;
  }
  if (!stat.isSymbolicLink()) {
    return safeRealPath(target) ?? target;
  }
  let linkTarget: string;
  try {
    linkTarget = fs.readlinkSync(target);
  } catch {
    return target;
  }
  return followSymlinkChain(path.resolve(path.dirname(target), linkTarget), depth + 1);
}

/**
 * Canonical comparison candidate for a gate verdict. The target file may not
 * exist yet (write creates it), so resolution goes through the PARENT
 * directory's realpath and re-attaches the basename — same TOCTOU strategy
 * as isPathInProject's fallback, no new semantics. Intermediate-directory
 * symlinks are resolved by the parent realpath; final-component symlinks
 * (including dangling ones) by followSymlinkChain.
 */
function resolveGateCandidate(filePath: string): string {
  const normalized = normalizeFilePath(filePath);
  const resolved = followSymlinkChain(normalized, 0);
  const realParent = safeRealPath(path.dirname(resolved));
  return realParent ? path.join(realParent, path.basename(resolved)) : resolved;
}

function isWithinRoot(root: string, candidate: string): boolean {
  // The root must be realpath-normalized too: on macOS a lexical
  // /var/folders/... root never contains a /private/var/folders/... candidate.
  const realRoot = safeRealPath(root) ?? path.resolve(root);
  const relative = path.relative(realRoot, candidate);
  return relative === "" || (!relative.startsWith("..") && !path.isAbsolute(relative));
}

function isPathInRoots(roots: readonly string[], candidate: string): boolean {
  return roots.some((root) => isWithinRoot(root, candidate));
}

File: deeporca-win/src/DeepOrca.Core/Common/PathBoundary.cs
using DeepOrca.Core.Types;

namespace DeepOrca.Core.Common;

// PathBoundary — 执行期路径边界门（对拍上游 common/path-boundary.ts，P0 沙盒 §4.1）。
// 无权限语义的纯路径原语：权限层分类 in/out-of-cwd，这里在每次 fs 触碰前强制。

public sealed record PathGrant
{
    /// <summary>允许的写根（realpath 归一）；恒含 realpath(projectRoot)。</summary>
    public required string[] WriteRoots { get; init; }
    /// <summary>允许的读根：realpath(projectRoot) + 豁免路径。</summary>
    public required string[] ReadRoots { get; init; }
    /// <summary>本次调用的 write-out-cwd 已解析为 allow（一次性授权无法表达为根列表）。</summary>
    public bool AllowWriteOutsideRoots { get; init; }
    /// <summary>本次调用的 read-out-cwd 已解析为 allow。</summary>
    public bool AllowReadOutsideRoots { get; init; }

    public static PathGrant ProjectOnly(string projectRoot) => new()
    {
        WriteRoots = [PathBoundary.ResolveGateRoot(projectRoot)],
        ReadRoots = [PathBoundary.ResolveGateRoot(projectRoot)],
    };
}

public sealed record GateVerdict
{
    public bool Ok { get; init; }
    public string? Reason { get; init; }
    public string? Scope { get; init; }

    public static GateVerdict Allowed => new() { Ok = true };
    public static GateVerdict Denied(string scope, string reason) => new() { Ok = false, Scope = scope, Reason = reason };
}

/// <summary>带隔离语义的 out-of-cwd 布尔推导（quarantined 工作区双 false）。</summary>
public static class PathBoundary
{
    public const int MaxSymlinkDepth = 10;

    public static (bool Write, bool Read) GrantOutsideRootsFlags(IEnumerable<string> scopes, bool quarantined)
    {
        if (quarantined) return (false, false);
        var set = scopes.ToHashSet();
        return (set.Contains("write-out-cwd"), set.Contains("read-out-cwd"));
    }

    /// <summary>读门（gateRead 直译）。无 grant 时退化为 projectRoot 独有、外溢双 false（fail-closed）。</summary>
    public static GateVerdict GateRead(PathGrant? grant, string filePath, string? projectRoot = null)
    {
        var candidate = ResolveGateCandidate(filePath);
        var roots = grant is { } g ? g.ReadRoots : projectRoot is { } r ? [r] : [];
        if (IsPathInRoots(roots, candidate)) return GateVerdict.Allowed;
        if (grant?.AllowReadOutsideRoots == true) return GateVerdict.Allowed;
        return GateVerdict.Denied("read-out-cwd",
            $"Read target is outside the allowed read boundary: {filePath}. This path was not authorized for out-of-project reads. If reading there is genuinely required, ask the user to grant the \"read-out-cwd\" permission.");
    }

    /// <summary>写门（gateWrite 直译）。</summary>
    public static GateVerdict GateWrite(PathGrant? grant, string filePath, string? projectRoot = null)
    {
        var candidate = ResolveGateCandidate(filePath);
        var roots = grant is { } g ? g.WriteRoots : projectRoot is { } r ? [r] : [];
        if (IsPathInRoots(roots, candidate)) return GateVerdict.Allowed;
        if (grant?.AllowWriteOutsideRoots == true) return GateVerdict.Allowed;
        return GateVerdict.Denied("write-out-cwd",
            $"Write target is outside the allowed write boundary: {filePath}. This path was not authorized for out-of-project writes. If writing there is genuinely required, ask the user to grant the \"write-out-cwd\" permission.");
    }

    /// <summary>门判定的规范化候选：链式符号链接 + 父目录 realpath（写目标可能尚不存在）。</summary>
    public static string ResolveGateCandidate(string filePath)
    {
        var resolved = FollowSymlinkChain(Path.GetFullPath(filePath), 0);
        var realParent = SafeRealPath(Path.GetDirectoryName(resolved) ?? resolved) ?? Path.GetDirectoryName(resolved) ?? resolved;
        return Path.Combine(realParent, Path.GetFileName(resolved));
    }

    /// <summary>授权根与门候选共享同一规范化器（否则 /tmp ↔ /private/tmp 确定性失配）。</summary>
    public static string ResolveGateRoot(string root) => ResolveGateCandidate(root);

    private static string FollowSymlinkChain(string target, int depth)
    {
        if (depth >= MaxSymlinkDepth) return target;
        FileSystemInfo info;
        try
        {
            info = new FileInfo(target);
            if (!info.Exists && !Directory.Exists(target))
            {
                // 不存在 → 追父链（写目标常见）
                var parent = Path.GetDirectoryName(target);
                return parent is null ? target : Path.Combine(FollowSymlinkChain(parent, depth + 1), Path.GetFileName(target));
            }
        }
        catch
        {
            return target;
        }

        var linkTarget = ResolveLink(target);
        if (linkTarget is null) return SafeRealPath(target) ?? target;
        return FollowSymlinkChain(Path.Combine(Path.GetDirectoryName(target) ?? "", linkTarget), depth + 1);
    }

    private static string? ResolveLink(string path)
    {
        try
        {
            return File.ResolveLinkTarget(path, returnFinalTarget: false)?.FullName
                ?? Directory.ResolveLinkTarget(path, returnFinalTarget: false)?.FullName;
        }
        catch
        {
            return null;
        }
    }

    public static string? SafeRealPath(string target)
    {
        try
        {
            var dir = Path.GetDirectoryName(target);
            var real = dir is null ? target : Path.Combine(Path.GetFullPath(dir), Path.GetFileName(target));
            if (File.Exists(target)) return Path.GetFullPath(target);
            if (Directory.Exists(real)) return real;
            return null;
        }
        catch
        {
            return null;
        }
    }

    private static bool IsPathInRoots(string[] roots, string candidate)
    {
        foreach (var root in roots)
        {
            var realRoot = SafeRealPath(root) ?? Path.GetFullPath(root);
            var relative = Path.GetRelativePath(realRoot, candidate);
            if (relative == "." || (relative != ".." && !relative.StartsWith("..") && !Path.IsPathRooted(relative)))
            {
                return true;
            }
        }
        return false;
    }
}